design_docs/support_external_swift_packages.md

```diff
@@ -0,0 +1,204 @@
+  with the external dependency information.
+- The `xcodeproj` rule will gain an `external_deps` attribute that expects targets that provide an
+  `ExternalSwiftPackagesInfo`.
+- The `xcodeproj` rule will generate the appropriate Xcode project entries to load the external
+  Swift package and associate it with the Xcode targets.
+
+### `xcodeproj_external_swift_packages` Rule and `ExternalSwiftPackagesInfo` Provider
+
+The mapping of an external Bazel target to Swift package dependency will occur in the
+`xcodeproj_external_swift_packages` rule. This rule provides an `ExternalSwiftPackagesInfo` instance
+that maps dependent package information to Bazel targets.
+
+
+#### Example: External Dependencies Downloaded using `http_archive`
+
+Let's assume that two external packages, `apple/swift-log` and `apple/swift-nio`, are used by a
+project. The `WORKSPACE` will contain declarations like the following:
+
+```python
+http_archive(
+    name = "apple_swift_log",
+    build_file = "@//bazel/third_party_deps:swift-log.BUILD",
+    sha256 = "3bc5e9691bb6f5426325d8a564f6dc72de05c5f05e6676a5e8745299aa07a09d",
+    strip_prefix = "swift-log-acbd69711395f5697b12beb4cf089929c9dbf3b4",
+    url = "https://github.com/apple/swift-nio/archive/acbd69711395f5697b12beb4cf089929c9dbf3b4.tar.gz",
+)
+
+http_archive(
+    name = "apple_swift_nio",
+    build_file = "@//bazel/third_party_deps:swift-nio.BUILD",
+    sha256 = "2bc5e9691bb6f5426325d8a564f6dc72de05c5f05e6676a5e8745299aa07a09c",
+    strip_prefix = "swift-nio-2.38.0",
+    url = "https://github.com/apple/swift-nio/archive/refs/tags/2.38.0.tar.gz",
+)
+```
+
+Let's also assume that the build files for these packages define Bazel targets that match the
+product names (e.g. the `Logging` product for `swift-log` is mapped to `@apple_swift_log//:Logging`.
+
+The `xcodeproj_external_swift_packages` and `xcodeproj` declarations would look like the following:
+
+```python
+xcodeproj_external_swift_packages(
+    name = "external_packages",
+    packages = [
+        # This example demonstrates downloading a Swift package at a particular commit.
+        xcodeproj_swift_package(
+            url = "https://github.com/apple/swift-log",
+            revision = "acbd69711395f5697b12beb4cf089929c9dbf3b4",
+            products_to_bazel_map = {
+                "Logging": "@apple_swift_log//:Logging",
+            }
+        ),
+        # This example demonstrates downloading a Swift package at a particular version tag.
+        xcodeproj_swift_package(
+            url = "https://github.com/apple/swift-nio",
+            exact_version = "2.38.0",
+            products_to_bazel_map = {
+                "NIO": "@apple_swift_nio//:NIO",
+                "NIOHTTP1": "@apple_swift_nio//:NIOHTTP1",
+            }
+        ),
+    ],
+)
+
+xcodeproj(
+    name = "xcodeproj",
+    project_name = "My App",
+    tags = ["manual"],
+    targets = XCODEPROJ_TARGETS,
+    external_deps = [
+        ":external_packages",
```

Comment:
So, from my understanding today we handle external packages with option 1 already. If we were to go with option 2, this would significantly increase the amount of work to adopt this rule for projects made up of only external `http_archive`s. Today, someone can simply drop a `xcodeproj` and the project should build just fine in Xcode. Instead, with this option the amount of rules required would be higher and we're basically duplicating the dependencies definitions once for building, and once for the IDE integration. Is my interpretation correct?

Reply:
In retrospect, I should not have proposed this as an either/or situation. In my mind, option 1 and option 2 offer different user experiences. (I wonder if there may be an Xcode performance benefit by reducing the number of targets in the project.) If I reframe the proposal as an alternative means to import external dependencies, I think that the extra mapping may be desirable for some projects, especially those with a large number of external transitive dependencies. 

In terms of the work adding the mapping for the direct dependencies for a project, it will vary. In general, the number of direct dependencies for a project will be lower than the number of transitive dependencies. In some cases, it will be much lower. For instance, there is [a Vapor application example in the rules_spm project](https://github.com/cgrindel/rules_spm/blob/main/examples/vapor/WORKSPACE#L43) that has 3 direct dependencies. When fully resolved, it results in 24 transitive dependencies. When writing this proposal, I was thinking about whether I would want to work with an Xcode project that has 5 targets for my project or 29 targets (5 source targets + 24 dependency targets). So, for this example, one would need to map 3 direct dependencies (i.e., ~25 lines of Starlark code) that would result in a project with 5 targets all of which are targets from the project. None of them would be for the external dependencies.

Back to the reframing, what if I update the design to reflect the following scenarios?

### Scenario 1: Initial Usage of `rules_xcodeproj`
- Add `xcodeproj` rule to your project and go.
- Everything works
- All of the transitive dependencies appear in Xcode.

### Scenario 2: Hide External Swift Packages
- Add `xcodeproj_external_swift_packages` mapping for the direct dependencies.
- Only the source targets appear in the resulting Xcode project.

Reply:
To me it sounds like this design doc should be focused on how to offer a native experience for projects using SPM through `rules_spm`. The fact that external Swift package are "hidden" is just a side-effect of using Xcode's native SPM functionality. If that's something that a project would like to offer as a usability or performance improvement, they should be able to do that for sure. For example, it could be another attribute named `uses_native_spm` in order to generate an Xcode project that uses native SPM when building with Xcode. The attribute would default to `False`, but setting that to `True` would open up the possibility of defining the native SPM to be added via the mapping of the dependencies as presented earlier. Does that make sense?

Reply:
I think it can go a little further: we can add SPM related values to `XcodeProjInfo`, but the default rules won't fill them in. This means we will default to Scenario 1. Then `rules_spm` (or any other custom rules) can supply `XcodeProjInfo` providers with the SPM values filled in. `rules_spm` could maybe have the attribute to allow users to choose if they would want SPM represented natively or not (though maybe it just always does it, since using `rules_spm` is already stating that it wants to "natively" integrate with SPM).